flake.nix

```
@@ -0,0 +1,205 @@
+          # developer environments. See https://devenv.sh/reference/options/ for a list
+          # of all possible options.
+          default = devenv.lib.mkShell {
+            inherit inputs pkgs;
+            modules = [
+              {
+                # Make use of the Starship command prompt when this development environment
+                # is manually activated (via `nix develop --impure`).
+                # See https://starship.rs/ for details on the prompt itself.
+                starship.enable = true;
+
+                # Configure packages to install.
+                # Search for package names at https://search.nixos.org/packages?channel=unstable
+                packages = with pkgs; [
+                  # Native dependencies for running Synapse.
+                  icu
+                  libffi
+                  libjpeg
+                  libpqxx
+                  libwebp
+                  libxml2
+                  libxslt
+                  sqlite
+
+                  # Native dependencies for unit tests (SyTest also requires OpenSSL).
+                  openssl
+
+                  # Native dependencies for running Complement.
+                  olm
+                ];
+
+                # Install Python and manage a virtualenv with Poetry.
+                languages.python.enable = true;
+                languages.python.poetry.enable = true;
+                # Automatically activate the poetry virtualenv upon entering the shell.
+                languages.python.poetry.activate.enable = true;
+                # Install all extra Python dependencies; this is needed to run the unit
+                # tests and utilitise all Synapse features.
+                languages.python.poetry.install.arguments = ["--extras all"];
+                # Install the 'matrix-synapse' package from the local checkout.
+                languages.python.poetry.install.installRootPackage = true;
+
+                # This is a work-around for NixOS systems. NixOS is special in
+                # that you can have multiple versions of packages installed at
+                # once, including your libc linker!
+                #
+                # Some binaries built for Linux expect those to be in a certain
+                # filepath, but that is not the case on NixOS. In that case, we
+                # force compiling those binaries locally instead.
+                env.POETRY_INSTALLER_NO_BINARY = "ruff";
+
+                # Install dependencies for the additional programming languages
+                # involved with Synapse development.
+                #
+                # * Rust is used for developing and running Synapse.
+                # * Golang is needed to run the Complement test suite.
+                # * Perl is needed to run the SyTest test suite.
+                languages.go.enable = true;
+                languages.rust.enable = true;
+                languages.rust.version = "latest";
```

I would be tempted to use the stable toolchain here

```suggestion
                languages.rust.version = "stable";
```

Our MSRV is a fairly old stable version and besides I don't really see why we would want to go for nightly in Synapse

---

Hmm, yes, and we should develop using a stable rustc backend lest we accidentally introduce nightly features when developing.

The reason I switched this to nightly was because some of our CI requiring using a nightly toolchain:

1. Running `cargo clippy --fix --benches --allow-dirty --allow-staged` will fail to fix errors in benchmark files on a stable toolchain due to ``#![feature]` may not be used on the stable release channel`. However, `--benches` are not checked by `./scripts-dev/lint.sh` and are rarely-touched files anyways, so this is probably fine to only check in CI.
2. Running `cargo-fmt` with a stable toolchain on Synapse results in `Warning: can't set `group_imports = StdExternalCrate`, unstable features are only available in nightly channel.`. This is printed out 4 times when running `./scripts-dev/lint.sh`. This warning is not the end of the world though, and we can probably live with it until we figure out some solution for it.

Nix does give us the power to install multiple rust toolchains at once, and we can probably engineer a solution that runs the linting against the nightly toolchain, whereas everything else uses the stable toolchain. Given that the only issue with using the stable toolchain right now is a minor warning, I'm happy to change it. Done in 44588aa69.

---

Hm, wasn't aware of the linting issues without nightly. It could make sense then but I suppose we can think on this later.